### ISSUE TITLE
Implement Curriculum Learning for Action Space Exploration

### ISSUE THREAD

**Issue by BY571** [TorchTrade/torchtrade#49]
## Problem

FuturesOneStepEnv has a **19-action space** that is too large for PPO to explore effectively with standard hyperparameters. Agents consistently converge to Action 6 (Long, SL=5%, TP=8%) without exploring alternatives.

**Evidence from research:**
- 10 iterations of reward function optimization exhausted
- Action 6 usage: 86-90% across all iterations
- Only ~2,000 exploration steps for 19 actions (need ~19,000)

---

## Proposed Solution: Curriculum Learning

**Gradually expand action space** during training to make exploration tractable.

### Phase-Based Curriculum

```python
class CurriculumScheduler:
    """Progressively unlock actions during training."""
    
    def __init__(self, total_frames: int):
        self.total_frames = total_frames
        self.phases = [
            # (frame_start, allowed_actions, description)
            (0, [0, 3, 6, 9, 12, 15, 18], "Phase 1: Core actions only (7 actions)"),
            (30_000, [0, 2, 3, 4, 6, 7, 9, 11, 12, 13, 15, 16, 18], "Phase 2: Add moderate variants (13 actions)"),
            (60_000, list(range(19)), "Phase 3: Full action space (19 actions)"),
        ]
    
    def get_allowed_actions(self, current_frame: int) -> List[int]:
        """Return list of allowed actions at current training frame."""
        for frame_threshold, actions, _ in reversed(self.phases):
            if current_frame >= frame_threshold:
                return actions
        return self.phases[0][1]  # Default to phase 1
```

### Integration with Environment

```python
class FuturesOneStepEnv(EnvBase):
    def __init__(self, ..., curriculum: Optional[CurriculumScheduler] = None):
        self.curriculum = curriculum
        self.current_frame = 0
    
    def step(self, tensordict):
        # Mask disallowed actions
        if self.curriculum:
            allowed = self.curriculum.get_allowed_actions(self.current_frame)
            action_mask = torch.zeros(len(self.action_map), dtype=torch.bool)
            action_mask[allowed] = True
            tensordict["action_mask"] = action_mask
        
        self.current_frame += 1
        return super().step(tensordict)
```

### Action Selection Strategy

**Phase 1 (0-30k frames): 7 Core Actions**
```
Action 0: Hold
Action 3: Long (SL=2%, TP=4%) - Conservative
Action 6: Long (SL=5%, TP=8%) - Moderate
Action 9: Long (SL=10%, TP=15%) - Aggressive
Action 12: Short (SL=2%, TP=4%) - Conservative
Action 15: Short (SL=5%, TP=8%) - Moderate
Action 18: Short (SL=10%, TP=15%) - Aggressive
```
- **Rationale**: Learn basic long/short strategies at different risk levels
- **Exploration budget**: 7 actions × 4,300 frames/action = adequate coverage

**Phase 2 (30k-60k frames): 13 Actions**
- Add intermediate SL/TP combinations
- Agent already understands long/short basics
- Can now learn fine-grained risk management

**Phase 3 (60k-100k frames): 19 Actions**
- Full action space unlocked
- Agent uses learned value function to guide exploration
- Fine-tuning SL/TP optimization

---

### Success Metrics

| Metric | Baseline | Curriculum Target |
|--------|----------|-------------------|
| Actions explored (>1% usage) | 2-3 | > 10 |
| Action entropy | 0.48 | > 1.2 |
| Action 6 dominance | 87% | < 40% |
| Short position usage | 0-5% | > 20% |
| Test Sharpe ratio | -5.77 | > -2.0 |

### Behavioral Improvements

1. **Better exploration**: Each phase allows thorough exploration before expansion
2. **Transfer learning**: Agent builds on learned strategies when new actions unlock
3. **Balanced strategies**: Forced to learn shorts early (not ignored like in baseline)
4. **Reduced premature convergence**: Can't lock into single action in early training

---

### Week 1: Core Implementation
- [ ] Create `CurriculumScheduler` in `torchtrade/training/curriculum.py`
- [ ] Add `action_mask` support to `FuturesOneStepEnv`
- [ ] Add `curriculum_config` to training configs
- [ ] Integrate with PPO actor (mask logits before softmax)

### Week 2: Testing
- [ ] Test Phase 1 only (7 actions, 100k frames)
- [ ] Test full 3-phase curriculum (30k/30k/40k split)
- [ ] Compare action distributions per phase
- [ ] Log curriculum phase transitions

### Week 3: Optimization
- [ ] Experiment with phase durations: [20k/30k/50k], [40k/40k/20k]
- [ ] Test different action groupings (risk-based vs R:R-based)
- [ ] A/B test curriculum vs baseline vs increased entropy

---

### Option A: Risk-Based Progression
```python
Phase 1: Only conservative actions (SL ≤ 3%)
Phase 2: Add moderate actions (SL ≤ 6%)
Phase 3: Add aggressive actions (SL > 6%)
```

### Option B: Directional Progression
```python
Phase 1: Long-only + hold
Phase 2: Add short positions
Phase 3: Full space
```

### Option C: Reward-To-Risk Progression
```python
Phase 1: High R:R ratio actions (TP/SL > 2.0)
Phase 2: Moderate R:R (TP/SL > 1.5)
Phase 3: All R:R ratios
```

---

## Configuration Example

```yaml
# config.yaml
curriculum:
  enabled: true
  strategy: "risk_based"  # or "phased", "directional"
  phases:
    - frame_start: 0
      max_actions: 7
      selection_criteria: "diverse_risk_levels"
    - frame_start: 30000
      max_actions: 13
    - frame_start: 60000
      max_actions: 19
  
  # Logging
  log_phase_transitions: true
  log_action_distribution_per_phase: true
```

---

## References

**Curriculum Learning:**
- Bengio et al. (2009) - "Curriculum Learning"
- Narvekar et al. (2020) - "Curriculum Learning for Reinforcement Learning Domains: A Framework and Survey"

**RL-Specific:**
- OpenAI Five (2018) - Used curriculum to master Dota 2 (180 hero combinations)
- AlphaStar (2019) - Curriculum over game difficulty and unit compositions

---

## Compatibility

- Should work with all FuturesOneStepEnv variants
- Compatible with existing PPO training pipeline
- Can combine with curiosity-driven exploration (issue #48)
- Should work with both one-step and sequential environments

---

## Related Issues

- #48 - Curiosity-driven exploration (complementary approach)
- Related files: `RewardFunctionResearch/FuturesOneStepEnv_Iteration5/reports/SUMMARY_iterations_5-10.md`
